### PR TITLE
[r8brain-free-src] add new port

### PR DESCRIPTION
**Pull request by CrSjimo**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Review by dg0yt — COMMENTED** [review]
The `CMakeLists.txt` should be just copied, for easy editing.

The CMake config must adopt the `unofficial` prefix pattern.

The source shares code which is in port pffft.

**Comment by CrSjimo** [comment]
> The `CMakeLists.txt` should be just copied, for easy editing.
> 
> The CMake config must adopt the `unofficial` prefix pattern.
> 
> The source shares code which is in port pffft.

@dg0yt What is your intention in mentioning "The source shares code which is in port pffft"? Should I use port pffft as a dependency instead of what is included in the upstream source code? But in that way, this port does not use the upstream source code as-is.

**Comment by Mengna-Li** [comment]
You should create a new CMakeList.txt file in the port/r8brain-free-src path.
Add the prefix 'unofficial' to the targets.

File: ports/r8brain-free-src/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    target_compile_definitions(r8bsrc PUBLIC R8B_PFFFT=1)
+endif()
+
+if(R8B_PFFFT_DOUBLE)
+    target_compile_definitions(r8bsrc PUBLIC R8B_PFFFT_DOUBLE=1)
+endif()
+
+if(R8B_INSTALL)
+    install(TARGETS r8bsrc
+        EXPORT ${PROJECT_NAME}Targets
```
Comment (Mengna-Li) [inline ports/r8brain-free-src/CMakeLists.txt:38]: Please add the prefix 'unofficial' to all your exports.

File: ports/r8brain-free-src/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        FILES ${_headers}
+        DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/r8brain
+    )
+    if(R8B_PFFFT_DOUBLE)
+        install(
+            FILES pffft_double/pffft_double.h
+            DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/r8brain/pffft_double
+        )
+    endif()
+endif()
```
Comment (Mengna-Li) [inline ports/r8brain-free-src/CMakeLists.txt:68]: Add a new line at the end of the file.

File: ports/r8brain-free-src/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,32 @@
+if(VCPKG_TARGET_IS_WINDOWS)
+    vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+endif()
+
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO avaneev/r8brain-free-src
+    REF version-${VERSION}
```
Comment (Mengna-Li) [inline ports/r8brain-free-src/portfile.cmake:8]:
```suggestion
    REF "version-${VERSION}"

```

File: ports/r8brain-free-src/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+endif()
+
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO avaneev/r8brain-free-src
+    REF version-${VERSION}
+    SHA512 ae2707aa76d3dc89153bbe755f134c497a6024d3bc06badbb078fe8bc5c7f09bfa277003c6915b341f35d86224930890abab89da0ebafb98722cf35f9a2222d9
+)
+
+file(COPY ${CMAKE_CURRENT_LIST_DIR}/CMakeLists.txt DESTINATION ${SOURCE_PATH})
```
Comment (Mengna-Li) [inline ports/r8brain-free-src/portfile.cmake:12]:
```suggestion
file(COPY "${CMAKE_CURRENT_LIST_DIR}/CMakeLists.txt" DESTINATION "${SOURCE_PATH}")

```

File: ports/r8brain-free-src/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        -DR8B_PFFFT=${R8B_PFFFT}
+        -DR8B_PFFFT_DOUBLE=${R8B_PFFFT_DOUBLE}
+)
+
+vcpkg_cmake_install()
+vcpkg_cmake_config_fixup(PACKAGE_NAME unofficial-r8brain CONFIG_PATH lib/cmake/unofficial-r8brain)
+vcpkg_copy_pdbs()
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
```
Comment (Mengna-Li) [inline ports/r8brain-free-src/portfile.cmake:32]:
```suggestion
vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
```

**Review by Mengna-Li — CHANGES_REQUESTED** [review]

File: ports/r8brain-free-src/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    FEATURES
+        pffft           R8B_PFFFT
+        pffft-double    R8B_PFFFT_DOUBLE
+)
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+        -DR8B_PFFFT=${R8B_PFFFT}
+        -DR8B_PFFFT_DOUBLE=${R8B_PFFFT_DOUBLE}
```
Comment (Mengna-Li) [inline ports/r8brain-free-src/portfile.cmake:24]:
```suggestion
    OPTIONS
        ${FEATURE_OPTIONS}
```

File: ports/r8brain-free-src/portfile.cmake [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+)
+
+vcpkg_cmake_install()
+vcpkg_cmake_config_fixup(PACKAGE_NAME unofficial-r8brain CONFIG_PATH lib/cmake/unofficial-r8brain)
+vcpkg_copy_pdbs()
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
```
Comment (Mengna-Li) [inline ports/r8brain-free-src/portfile.cmake:32]: Please add a new line at the end of the file.

**Review by Mengna-Li — CHANGES_REQUESTED** [review]

**Comment by Mengna-Li** [comment]
@CrSjimo  Usage test failed: could not find header files in my test project.
Repro steps:
1. Create a CMake project.
2. Add provided usage in CMakeLists.txt.
3. Add include<r8brain/r8bconf.h>

**Comment by Mengna-Li** [comment]
Usage test pass on x64-windows.

**Review by Mengna-Li — APPROVED** [review]

File: ports/r8brain-free-src/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    )
+
+    install(EXPORT unofficial-${PROJECT_NAME}Targets
+        FILE "unofficial-${PROJECT_NAME}Config.cmake"
+        NAMESPACE unofficial::r8brain::
+        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/unofficial-${PROJECT_NAME}
+    )
+
+    file(GLOB _headers *.h *.inc)
+    list(FILTER _headers EXCLUDE REGEX r8butil.h)
```
Comment (BillyONeal) [inline ports/r8brain-free-src/CMakeLists.txt:54]: There should probably be a comment here describing why this one is special?

File: ports/r8brain-free-src/portfile.cmake [hunk 9]
```diff
@@ -0,0 +1,31 @@
+if(VCPKG_TARGET_IS_WINDOWS)
+    vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+endif()
```
Comment (BillyONeal) [inline ports/r8brain-free-src/portfile.cmake:3]: Upstream explicitly documents use as a DLL, is there a particular reason that is blocked? ( https://github.com/avaneev/r8brain-free-src?tab=readme-ov-file#dynamic-link-library )

**Review by BillyONeal — CHANGES_REQUESTED** [review]
Hi there, thanks for the new port! I'm sorry that I have a bunch of things that I think need to be addressed before we can merge this.

1. It looks like the license is wrong and that there are vendored dependencies in here. https://learn.microsoft.com/vcpkg/contributing/maintainer-guide#do-not-use-vendored-dependencies  
  From https://github.com/avaneev/r8brain-free-src
    >If you do not have access to the Intel IPP then you may consider enabling the PFFFT which is only slightly slower than Intel IPP FFT in performance. There are two macros available: R8B_PFFFT and R8B_PFFFT_DOUBLE. The first macro enables PFFFT that works in single-precision resolution, thus limiting the overall resampler's precision to 24-bit sample rate conversions (for mission-critical professional audio applications, using the R8B_PFFFT macro is not recommended as its peak error is quite large). The second macro enables PFFFT implementation that works in double-precision resolution, making use of SSE2, AVX, and NEON intrinsics, yielding precision that is equal to both Intel IPP and Ooura FFT implementations.
    > [...]
    > r8brain-free-src is bundled with the following code:
    > 
    > FFT routines Copyright (c) 1996-2001 Takuya OOURA. [Homepage](http://www.kurims.kyoto-u.ac.jp/~ooura/fft.html)
    > PFFFT Copyright (c) 2013 Julien Pommier. [Homepage](https://bitbucket.org/jpommier/pffft)
    > PFFFT DOUBLE Copyright (c) 2020 Hayati Ayguen, Dario Mambro. [Homepage](https://github.com/marton78/pffft)
    
    MIT does not seem like the correct license given that these parts are very much not MIT.
2. The name `r8brain-free-src` indeed matches the name of the repo, but your replacement build system just calls it `r8brain`. It seems like either we should include the `-free-src` part everywhere or nowhere? Unfortunately `r8brain` does not result in this being first in typical search results... Perhaps the port name should be `avaneev-r8brain`? I observe Repology knows this port via ConanCenter: https://repology.org/project/r8brain-free-src/versions
3. It is difficult to evaluate whether this port is correct because it is replacing the build system entirely, and upstream does not document any particular expected usage pattern. The closest thing I was able to find to examples use relative paths that are only valid within the upstream repo itself, e.g. https://github.com/avaneev/r8brain-free-src/blob/master/bench/masstest.cpp . I think we need upstream usage instructions that let us confirm that the build system replacement does what people expect the resulting build system bindings to do.  
  To be clear, what I mean by usage instructions is things like 'What are downstream consumers expected to #include'?
4. There are features in this port that don't appear to be wired up correctly; for example https://github.com/avaneev/r8brain-free-src/blob/master/r8bconf.h has macros inside that if downstream customers don't exactly match, they get silent bad codegen, if I understand correctly. I think the results of the features the vcpkg user has selected should be baked into the resulting headers so that customers can't make that mistake. (We might normally merge a port with an upstream build system that requires macros exported via CMake bindings to work, but given that we are replacing the build system entirely here, it should work for all vcpkg customers, including those customers who can't use CMake bindings)

**Comment by CrSjimo** [comment]
@BillyONeal  Thanks for your suggestions. 

1. I have noticed the vendored dependencies that r8brain-free-src contains, but I have some trouble dealing with them. According to r8brain-free-src's readme, there are 3 bundled dependencies:

    > FFT routines Copyright (c) 1996-2001 Takuya OOURA. [Homepage](http://www.kurims.kyoto-u.ac.jp/~ooura/fft.html)
    > PFFFT Copyright (c) 2013 Julien Pommier. [Homepage](https://bitbucket.org/jpommier/pffft)
    > PFFFT DOUBLE Copyright (c) 2020 Hayati Ayguen, Dario Mambro. [Homepage](https://github.com/marton78/pffft)
    
    After checking the source code, I found that the author of r8brain-free-src seems to have modified the codes of these dependencies, making them different from their upstream. I also found that the licenses of pffft and pffft-double are not compatible with MIT. Therefore, should I list their licenses in the `copyright` file?

2. The upstream provides neither any useful information about building nor any build scripts. As is mentioned in avaneev/r8brain-free-src#15 and avaneev/r8brain-free-src#20, the author of r8brain-free-src just said

    > There's little to "build" in this library, you basically need to compile and include a single cpp file, the library is "inline" for the most part. So, I do not see a need in CMake build.
    
    Also, it seems that the author is not going to add any official CMake build scripts for this library for some time to come. So I have no idea of the usage of headers.
    
    In my CMakeLists, I just simply put all source files into the library. `r8butil.h` is only included by utility programs and not used by the library code, so I excluded it. The DLL provided by the upstream is a [C API wrapper](https://github.com/avaneev/r8brain-free-src/blob/master/DLL/r8bsrc.h) of the library. The library code itself does not export any symbols, so I decided to set `ONLY_STATIC_LIBRARY`.

3. After checking the feature macros, I found that they are only used in header files to determine which FFT backend to use, and they are not used in the source file (if I comprehend its logic correctly). Instead of using vcpkg to select features, perhaps I should compile source codes of all these FFT backends, and the user could define these macros directly when including the library header file.

I feel there are many uncertainties about the upstream. I am considering asking the upstream author about these things.

File: ports/r8brain-free-src/pffft-double-license [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+THIS SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING, BUT NOT LIMITED TO THE WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE CONTRIBUTORS OR COPYRIGHT
+HOLDERS BE LIABLE FOR ANY CLAIM, INDIRECT, INCIDENTAL, SPECIAL,
+EXEMPLARY, OR CONSEQUENTIAL DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE
+SOFTWARE.
```
Comment (Mengna-Li) [inline ports/r8brain-free-src/pffft-double-license:36]: Please add a new line at the end of the file.

File: ports/r8brain-free-src/pffft-license [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+THIS SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING, BUT NOT LIMITED TO THE WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE CONTRIBUTORS OR COPYRIGHT
+HOLDERS BE LIABLE FOR ANY CLAIM, INDIRECT, INCIDENTAL, SPECIAL,
+EXEMPLARY, OR CONSEQUENTIAL DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE
+SOFTWARE.
```
Comment (Mengna-Li) [inline ports/r8brain-free-src/pffft-license:45]: Add a new line.

**Review by Mengna-Li — CHANGES_REQUESTED** [review]

**Review by Mengna-Li — APPROVED** [review]

**Comment by JavierMatosD** [comment]
Thank you for your contribution! After reviewing this PR, we unfortunately cannot accept it due to the inclusion of vendored dependencies that have been modified.

Including modified vendored dependencies conflicts with vcpkg’s policy of promoting a consistent and centralized dependency management ecosystem. Vendoring introduces maintenance challenges, makes it harder for downstream consumers to understand the provenance of the dependencies, and complicates updates to those dependencies in the future. e.g., If we accept this PR as-is, we lose the ability to add `PFFFT` or `PFFFT-Double` as standalone libraries in the future, which would be a significant limitation for other users of vcpkg.

If this library requires the modified dependencies, we encourage you to use a [custom vcpkg registry](https://learn.microsoft.com/en-us/vcpkg/maintainers/registries) or an [overlay port](https://learn.microsoft.com/en-us/vcpkg/concepts/overlay-ports) to maintain these modifications separately. This approach allows you to continue using the modified versions locally without introducing them into the core vcpkg repository.

Please refer to our [policy on vendored dependencies](https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#do-not-use-vendored-dependencies) for more details on why we have this restriction.

**Pull request closed without merging**